auto_submit/test/service/validation_service_test.dart

```diff
@@ -99,14 +105,103 @@ void main() {
+    late PullRequestHelper flutterRequest;
+    gh.RepositorySlug slug = gh.RepositorySlug('flutter', 'cocoon');
+    final gh.PullRequest pullRequest = generatePullRequest(prNumber: 0, repoName: slug.name);
+    late qr.QueryResult queryResult;
+    late FakeValidationService validationService;
+    final FakePubSub pubsub = FakePubSub();
+
+    setUp(() {
+      flutterRequest = PullRequestHelper(
+        prNumber: 0,
+        lastCommitHash: oid,
+        reviews: <PullRequestReviewHelper>[],
+      );
+      githubService.checkRunsData = checkRunsMock;
+      githubService.createCommentData = createCommentMock;
+      //  final FakePubSub pubsub = FakePubSub();
+      queryResult = createQueryResult(flutterRequest);
+      validationService = FakeValidationService(config);
+    });
+
+    test('ProcessMerge is successful.', () async {
+      validationService.processMergeReturn = true;
+      bool validated = await validationService.processMergeSafely(
+          config: config,
+          gitHubService: githubService,
+          messagePullRequest: pullRequest,
+          queryResult: queryResult,
+          pubSub: pubsub,
+          ackId: 'id',
+          repositorySlug: pullRequest.base!.repo!.slug(),
+          prNumber: 0,
+          prLabel: 'merge');
+
+      expect(validated, isTrue);
+      expect(validationService.threwException, isFalse);
+    });
+
+    test('ProcessMerge is unsuccessful.', () async {
+      bool validated = await validationService.processMergeSafely(
+          config: config,
+          gitHubService: githubService,
+          messagePullRequest: pullRequest,
+          queryResult: queryResult,
+          pubSub: pubsub,
+          ackId: 'id',
+          repositorySlug: pullRequest.base!.repo!.slug(),
+          prNumber: 0,
+          prLabel: 'merge');
+
+      expect(validated, isFalse);
+      expect(validationService.threwException, isFalse);
+    });
+
+    test('ProcessMerge is unsuccessful.', () async {
+      validationService.returnValue = false;
+      bool validated = await validationService.processMergeSafely(
+          config: config,
+          gitHubService: githubService,
+          messagePullRequest: pullRequest,
+          queryResult: queryResult,
+          pubSub: pubsub,
+          ackId: 'id',
+          repositorySlug: pullRequest.base!.repo!.slug(),
+          prNumber: 0,
+          prLabel: 'merge');
+
+      expect(validated, isFalse);
+      expect(validationService.threwException, isTrue);
+    });
+  });
+}
+
+class FakeValidationService extends ValidationService {
+  FakeValidationService(super.config);
+
+  bool processMergeReturn = false;
+  bool returnValue = true;
+  bool threwException = false;
+
+  @override
+  Future<bool> processMerge(Config config, qr.QueryResult queryResult, gh.PullRequest messagePullRequest) async {
+    if (returnValue) {
+      return processMergeReturn;
+    }
+    threwException = true;
+    throw Exception();
+  }
```

Comment:
Instead of faking this service, we should validate the logics. Other tests in this file may give you some hint.